python/ray/workflow/http_event_provider.py

```diff
@@ -0,0 +1,240 @@
+        if event_key in self.event_key_payload[workflow_id]:
+            raise WorkflowEventHandleError(workflow_id, f"The same {event_key} is used to get payload again.")
+        self.event_key_payload[workflow_id][event_key] = asyncio.Future()
+        return await self.event_key_payload[workflow_id][event_key]
+
+    async def report_checkpointed(self, event, confirmation):
+        self.payload_record[event].set_result(confirmation)
+
+    async def report_checkpointed_eca(self, workflow_id, event_key, confirmation):
+        self.event_key_checkpointed[workflow_id][event_key].set_result(confirmation)
+
+
+class HTTPListener(EventListener):
+    """HTTPLister is defined to work with the HTTPEventProvider. It implements two
+    APIs, poll_for_event() and event_checkpointed(). An instance of HTTPListener will
+    be started by the get_message() of the workflow.wait_for_event() to listen for an event
+    from the HTTPEventProvider instance (a Ray Serve deployment). Another instance of HTTPListener
+    will be started by the message_committed() of the workflow.wait_for_event() to confirm
+    that the event has been checkpointed.
+
+
+    Example definition
+    ==================
+
+    ```
+    class HTTPListener:
+
+        def __init__(self):
+            self.handle = ray.serve.get_deployment().get_handle()
+
+        async def poll_for_event(self, *args, **kwargs) -> Any:
+
+        async def event_checkpointed(self, event) -> None:
+
+    ```
+
+    Example Usage
+    =============
+    >>> from ray.workflow.http_event_provider import HTTPEventProvider, HTTPListener
+    >>> ray.init(address='auto', namespace='serve')
+    >>> serve.start(detached=True)
+    >>> HTTPEventProvider.deploy()
+    >>> event_step = workflow.wait_for_event( # doctest: +SKIP
+    ...     HTTPListener, workflow_id='', event_key='')
+    >>> handle_event = ... # doctest: +SKIP
+    >>> handle_event.step(event_step).run() # doctest: +SKIP
+    >>>
+
+    """
+    def __init__(self):
+        from ray import serve
+        self.handle = ray.serve.get_deployment('myEventProvider').get_handle(sync=True)
+
+    async def poll_for_event(self, *args, **kwargs) -> Event:
+        workflow_id = kwargs.pop("workflow_id", None)
+        event_key = kwargs.pop("event_key", None)
+        if workflow_id is None or event_key is None:
+            if workflow_id is None:
+                workflow_id = 'None'
+            raise WorkflowEventHandleError(workflow_id, \
+                f"poll_for_event() needs both kwargs workflow_id='' and event_key=''")
+        payload = await self.handle.get_event_payload.remote(workflow_id, event_key)
+        return payload
+
+    async def event_checkpointed(self, event: Event) -> None:
+        """Called after an event has been checkpointed and a transaction can
+        be safely committed."""
+        await self.handle.report_checkpointed.remote(event, True)
+
+class HTTPListenerInEventCoordinator(EventListener):
```

Comment:
Remove this one in this PR. Also I'm not sure whether we need this or not. Let's see whether we can just update the HttpListener to support this.

Reply:
Done